### ISSUE TITLE
Making Grav stateless

### ISSUE THREAD

**Issue by jlehtinen** [getgrav/grav#2215]
_In reference to this post by @rhukster in https://github.com/getgrav/grav/issues/1767#issuecomment-426647637_ — (I'm relaying this message from our backend devs, so bear with me with any inaccuracies in terms etc!)

For whatever reason (probably not Grav related), sometimes PHP, Nginx, or AWS instances can just randomly break or die, which is the main reason why we've been trying to use two identical nodes running behind a load balancer. So, if one of the instances fails, the system automatically switches to the working node.

However, at the moment Grav doesn't support stateless server instances, that is, cache file naming (mostly images) will vary between otherwise identical server nodes. This could be solved by using session persistence, which ensures files will only be loaded from one of the nodes for the length of that session, but this is somewhat archaic and also breaks some things, like CDN's _whole site acceleration_ we'd like to use. The trend, apparently, has been to keep the session state outside the server for years now.

So, in short, while we've been moving mountains to create a system where everything is robust, fault-tolerant, and auto healing, but unfortunately, there's still a single node that acts as a gateway between this robust system and the world that doesn't yet support this in a satisfactory way.

Do you think this could be solved using some method of file naming that's not dependent on outside circumstances (server the instance is running on)? Possibly using hashes of the file itself, and maybe incorporating necessary file parameters. In our (developers') view, this would allow Grav to scale better — and make it a _truly_ flat file, stateless platform, removed from "the application server territory" (Wordpress…)

Thanks for any thoughts on this!

**Comment by rhukster** [comment]
Thanks for the info @jlehtinen i'm going to discuss this with @mahagr to see what we can do to improve this.

**Comment by mahagr** [comment]
@jlehtinen I need some clarifications on how your system is set up. So you have two servers behind a load balancer, where both of them have an identical installation of Grav, but are independent -- meaning that they have both local copies of the files?

In Grav, images are dynamically created on request, which means that the image only exists in a single node or you need to use CDN to serve the images (or have a way to get a copy of the file from the proper server after it has been generated). I think the easiest way to fix this is to create a reverse proxy setup to load the missing images from the other server.

**Comment by jlehtinen** [comment]
Thank you @rhukster and @mahagr for looking into this, really appreciate it. I'll forward the question to our devs and let you know!

**Comment by mahagr** [comment]
Jään odottelemaan vastaustasi! (I'll be waiting for the response)

**Comment by jlehtinen** [comment]
Mahtavaa! Awesome! 😃

**Comment by skyscooby** [comment]
Has there been any progress on this front? Evaluating this product for a client but not being able to load balance it across many server nodes is a non-starter.. 

Perhaps this issue could be renamed to Load balancing support or product scale out.
